### ISSUE TITLE
Reverse `tree_order`, but with children last

### ISSUE THREAD

**Issue by ge0rg** [feincms/django-tree-queries#6]
I'm working on a kind-of-a-web-forum implementation, and I'd like to return a list sorted by the published date (newest fist), and additionally with sticky posts at the beginning of the list, though I can work around the "sticky posts first" issue by doing two queries, filtered by `sticky=True/False`.

The tree query is only using the first `Meta.ordering` column, and it's not possible to negate it in `Meta.ordering`:

```
django.db.utils.OperationalError: near "-": syntax error
```

Therefore I've set up `Meta.ordering = ('published_date', )` to get into the `tree_ordering` array, and then order the response like this:

```python
ForumPost.objects.with_tree_fields() \
    .filter(is_sticky=False) \
    .extra(order_by=['-__tree.tree_ordering', '__tree.tree_depth'])
```

However, the reverse ordering on `tree_ordering` makes the children be returned first, and `tree_depth` isn't accounted for because shorter arrays are "less" than longer arrays. I can't sort the result by `tree_depth` nor by `published_date` because that would separate the children from their parents.

I intend to paginate the query, so separating families is not an option. Is there any other way to reverse-sort on the `tree_ordering` column, but to keep longer array values after shorter ones?

**Comment by matthiask** [comment]
Hi,

Maybe there is a way to "generate" an ascending ordering for nodes, e.g. instead of storing timestamps since 1970 store "seconds until 2100" or something. It sounds hacky but it may solve the problem since later posts will have lower "seconds until 2100" values which makes BFS do the right thing?

**Comment by ge0rg** [comment]
Yes, that's what I have considered doing indeed. Also add a huge constant value for the stickyness and they'll get pushed to the front. I just hoped there would be a cleaner solution ;-)

**Comment by matthiask** [comment]
There certainly is a cleaner solution, but it may require custom SQL or a different query compiler :-/

django-cte is a bit more generic and may offer a way to achieve your goals without custom SQL but I don't know – I've never used it in a project. I sort of like the simplicity of django-tree-queries' implementation, relatively speaking.

**Comment by matthiask** [comment]
I thought about this some more because it came up again in #60. I want to document my thinking and why this is slightly harder than it may look at first sight because I feel as if I have already gone down that road once and don't want to redo it over and over.

The DFS ordering works by building arrays of each node's `ordering` value and applying an ascending ordering. The ascending ordering has the nice property that it 1. sorts siblings correctly and 2. also sorts parents before their children (since an array which is a short prefix of another is "less" than the other). Reversing the ordering would easily solve 1. but wouldn't keep 2. and I don't immediately see a quick way to mitigate that.

```
# select '{1}'::int[] < '{1,2}'::int[];
 ?column?
----------
 t
(1 row)
```

**Comment by matthiask** [comment]
Yes, that's an interesting idea. I think you could even use `GeneratedField(F("position") * -1, db_persist=True)` for this, and django-tree-queries wouldn't have to change? https://docs.djangoproject.com/en/dev/ref/models/fields/#generatedfield

(Not changing is an advantage insofar as we can keep the complexity of handling various field types in the expression code instead of reimplementing it partially and worse in django-tree-queries.)

**Comment by rhomboss** [comment]
The GeneratedField is an good solution if you are concerned about data duplication. I've done something similar by overriding the save method of one of my models so that it updates a field with the inverse of another.

The field types are an issue. One way I envision handling it is by performing a regular desc ordering query and then numbering the result with ascending integers and then performing a tree query that orders siblings on the new field. This method would also allow you to order based on multiple fields. However it would essentially be turning one query into two which might be undesirable.

**Comment by rhomboss** [comment]
I made the following quick edit to `tree_ordering` in the django-tree-queries `SQLCompiler` to test my window function idea and it seems to work.
```
CTE_POSTGRESQL_WITH_INTEGER_ORDERING = """
    WITH RECURSIVE __tree (
        "tree_depth",
        "tree_path",
        "tree_ordering",
        "tree_pk"
    ) AS (
        SELECT
            0 AS tree_depth,
            array[T.{pk}] AS tree_path,
            array[ROW_NUMBER() OVER(ORDER BY {order_by})] AS tree_ordering,
            T."{pk}"
        FROM {db_table} T
        WHERE T."{parent}" IS NULL

        UNION ALL

        SELECT
            __tree.tree_depth + 1 AS tree_depth,
            __tree.tree_path || T.{pk},
            __tree.tree_ordering || ROW_NUMBER() OVER(ORDER BY {order_by}),
            T."{pk}"
        FROM {db_table} T
        JOIN __tree ON T."{parent}" = __tree.tree_pk
    )
    """
```

This edit requires a raw SQL ordering string to be put in `.order_siblings_by()` since the api does not properly interpret 
`-some_field` to `some_field DESC` _yet_.

I've only tested this on descending fields but I don't see a reason for it to not also work with ordering over multiple fields. I'm less sure about aggregates and annotations. Overall I think this might be a good way of expanding the usability of 
django-tree-queries without increasing its complexity.

**Comment by matthiask** [comment]
Thanks! That's definitely an interesting approach. Do you know if something similar also exists for MariaDB/MySQL and sqlite?

**Comment by rhomboss** [comment]
According to their documentation [MariaDB](https://mariadb.com/kb/en/row_number/), [MySQL](https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html#function_row-number), and [sqlite](https://www.sqlite.org/windowfunctions.html#introduction_to_window_functions) all support window functions and more specifically `ROW_NUMBER`. Truth be told I'm not that familiar with SQL though and I don't have anything other than Postgres setup on my machine so I'm not in a position to test the other databases.

Also I'd like to point out that my method does make it so that the `tree_ordering` field is always the ordinal series 1,2,3,...etc. I'm not sure if that really matters, but I thought I should point it out.

**Comment by matthiask** [comment]
> Also I'd like to point out that my method does make it so that the tree_ordering field is always the ordinal series 1,2,3,...etc. I'm not sure if that really matters, but I thought I should point it out.

Yeah, that shouldn't matter. `tree_path` is used e.g. for `ancestors()`, but `tree_ordering` is only ever used for the sort in the database.

**Comment by rhomboss** [comment]
After digging around in the django sql backend I found that django supports both [window functions](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#window-functions) and [ROW_NUMBER()](https://docs.djangoproject.com/en/3.2/ref/models/database-functions/#rownumber) natively. Though this support only seems to be intended for annotations and aggregations I was able to use it to add the following method to the `TreeCompiler` which will generate a ROW_NUMBER() expression that can be plugged in place of the order_by field.

```
from django.db.models.expressions import Window
from django.db.models.functions.window import RowNumber

...

def sibling_order_sql(self):
        """
            Creates a raw sql ROW_NUMBER() window expression that can be used
            for ordering
        """
        opts = self.query.model._meta

        # Create a list of OrderBy objects that can be complied in a window function
        order_by_objs = []
        for field in self.query.sibling_order:
            # Alias must be 't' to match the one hardcodded in the CTE even though
            # the hardcoded one is capital 'T'
            for order_obj, is_ref in self.find_ordering_name(field, opts, alias='t'):
                order_by_objs.append(order_obj)

        row_window = Window(expression=RowNumber(), order_by=order_by_objs)
        order_sql, order_params = self.compile(row_window)

        return order_sql
```

There are a couple things to note with this implementation:
1. For some reason the `db_table` alias is `t` instead of `T` even though it looks like `T` in the hardcoded CTE strings. I don't know if this is just an issue with PostgreSQL or if its true for all databases.
2. While I assume that using the native django `Window` and `RowNumber` classes means that this method works for all django supported databases I have not tested anything other than PostgreSQL.
3. While this method does add support for ordering in any direction and ordering over multiple fields it does not add support for ordering on related fields. Ordering on related fields would require adding support for `JOIN`s in the CTE strings which is work I leave for someone in the future.

Also, does ordering in this way remove the need for separate text and integer ordering CTE strings since `ROW_NUMBER()` converts everything into ascending integers?

**Comment by rhomboss** [comment]
I got around to testing the `ROW_NUMBER()` solution on SQLite over the weekend and discovered that it wouldn't work as is. It turns out that, unlike PostgreSQL, SQLite and MySQL do not support placing aggregate or window functions in the second select statement of a recursive CTE. To get around that limitation I moved the `ROW_NUMBER()` function to a CTE of its own.
```
CTE_SQLITE3_WITH_INTEGER_ORDERING = """
    WITH RECURSIVE __rank_table({pk}, {parent}, rank_order) AS (
        SELECT
            {pk},
            {parent},
            row_number() OVER (ORDER BY {order_by})
        FROM {db_table}
        # Adding JOINs here will probably give access to ordering on related fields as well
    ),
    __tree(tree_depth, tree_path, tree_ordering, tree_pk) AS (
        SELECT
            0 tree_depth,
            printf("{sep}%%s{sep}", {pk}) tree_path,
            printf("{sep}%%020s{sep}", T.rank_order) tree_ordering,
            T."{pk}" tree_pk
        FROM __rank_table T # Note: we are using the __rank_table as the base table now
        WHERE T."{parent}" IS NULL

        UNION ALL

        SELECT
            __tree.tree_depth + 1,
            __tree.tree_path || printf("%%s{sep}", T.{pk}),
            __tree.tree_ordering || printf("%%020s{sep}", T.rank_order),
            T."{pk}"
        FROM __rank_table T # Note: we are using the __rank_table as the base table now
        JOIN __tree ON T."{parent}" = __tree.tree_pk
    )
    """
``` 
Basically there is now an extra non-recursive CTE called `__rank_table` that fetches all the data needed by `__tree` including the extra column `rank_order` which contains the results of `ROW_NUMBER()`.

I'm still doing a bit of testing but this method of ordering already passes everything in the `testapp` except `test_attributes` but that's to be expected since `tree_ordering` has changed.

**Comment by matthiask** [comment]
Fixed by #62
